### PR TITLE
Add color-scheme article

### LINKED ISSUE #2265: content: color-scheme [Mar 31]
Deadline: 2020 Mar 31 (Soft)

### PR DESCRIPTION
**Pull request by tomayac**
Fixes #2265

Changes proposed in this pull request:

- Adds the article

### CODE REVIEW

**Comment by tomayac** [comment]
@lilles FYI, the `color-scheme` article is ready for review. If you can, please take a look, here's the [staging preview deeplink](https://deploy-preview-2481--web-dev-staging.netlify.com/color-scheme/).

File: src/site/content/en/blog/color-scheme/index.md [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+{% endAside %}
+
+## Interplay with `prefers-color-scheme`
+
+The interplay of the `color-scheme` CSS property and the corresponding meta tag
+with the `prefers-color-scheme` user preference media feature may seem confusing at first.
+In fact, they play together really well.
+The most important thing to understand is that `color-scheme`
+exclusively determines the default appearance,
+whereas `prefers-color-scheme` also determines the stylable appearance.
```
Comment (jpmedley) [inline src/site/content/en/blog/color-scheme/index.md]:
```suggestion
whereas `prefers-color-scheme` determines the stylable appearance.
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+    <fieldset>
+      <legend>Lorem ipsum</legend>
+      <button type="button">Lorem ipsum</button>
+    </fieldset>
+  </form>
+</body>
+```
+
+The inline CSS code on the page
+sets the `<fieldset>`'s `background-color` to `gainsboro` in the general case,
```
Comment (jpmedley) [inline src/site/content/en/blog/color-scheme/index.md]:
```suggestion
sets the `<fieldset>` element's `background-color` to `gainsboro` in the general case,
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+Via the `<meta name="color-scheme" content="dark light">` element,
+the page tells the browser that it supports a dark and a light theme,
+with a preference for a dark theme.
+
+Depending on whether the operating system is set to dark or light mode,
+the whole page appears light on dark, or vice versa, based on the user-agent stylesheet.
+There is *no* additional developer-provided CSS involved to change the paragraph text
+or the background color of the page.
+
+Note how the `<fieldset>`'s `background-color` changes
```
Comment (jpmedley) [inline src/site/content/en/blog/color-scheme/index.md]:
```suggestion
Note how the `<fieldset>` element's `background-color` changes
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+based on whether dark mode is enabled or not, following the rules
```
Comment (jpmedley) [inline src/site/content/en/blog/color-scheme/index.md]:
```suggestion
based on whether dark mode is enabled, following the rules
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+in the developer-provided inline stylesheet on the page.
+It is either `gainsboro` or `darkslategray`.
+
+<figure>
+  <img src="light-styles.png" width="3440" height=" 1386">
+  <figcaption>
+    <strong>Light mode:</strong> Styles specified by the developer and the user-agent.
+    The text is black and the background is white as per the user-agent stylesheet.
+    The <code>&lt;fieldset&gt;</code>'s <code>background-color</code> is <code>gainsboro</code>
```
Comment (jpmedley) [inline src/site/content/en/blog/color-scheme/index.md]:
```suggestion
    The <code>&lt;fieldset&gt;</code> element's <code>background-color</code> is <code>gainsboro</code>
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+    as per the inlined developer stylesheet.
+  </figcaption>
+</figure>
+
+<figure>
+  <img src="dark-styles.png" width="3440" height=" 1386">
+  <figcaption>
+    <strong>Dark mode:</strong> Styles specified by the developer and the user-agent.
+    The text is white and the background is black as per the user-agent stylesheet.
+    The <code>&lt;fieldset&gt;</code>'s <code>background-color</code> is <code>darkslategray</code>
```
Comment (jpmedley) [inline src/site/content/en/blog/color-scheme/index.md]:
```suggestion
    The <code>&lt;fieldset&gt;</code> element's <code>background-color</code> is <code>darkslategray</code>
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+    as per the inlined developer stylesheet.
+  </figcaption>
+</figure>
+
+The `<button>`'s appearance is controlled by the user-agent stylesheet.
```
Comment (jpmedley) [inline src/site/content/en/blog/color-scheme/index.md]:
```suggestion
The `<button>` element's appearance is controlled by the user-agent stylesheet.
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+Its `color` is set to the
+[`ButtonText`](https://drafts.csswg.org/css-color/#valdef-system-color-buttontext)
+system color, and its `background-color` and the four `border-color`s are set to the system color
+[`ButtonFace`](https://drafts.csswg.org/css-color/#valdef-system-color-buttonface)
+
+<figure>
+  <img src="light-buttonface.png" width="3440" height=" 1386">
+  <figcaption>
+    <strong>Light mode:</strong> The <code>background-color</code> and the various
+    <code>border-color</code>s are set the the <a href="https://drafts.csswg.org/css-color/#valdef-system-color-buttonface">ButtonFace</a>
```
Comment (jpmedley) [inline src/site/content/en/blog/color-scheme/index.md]:
```suggestion
    <code>border-color</code>s are set to the <a href="https://drafts.csswg.org/css-color/#valdef-system-color-buttonface">ButtonFace</a>
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+    system color.
+  </figcaption>
+</figure>
+
+Now note how the `<button>`'s `border-color` changes.
```
Comment (jpmedley) [inline src/site/content/en/blog/color-scheme/index.md]:
```suggestion
Now note how the `<button>` element's `border-color` changes.
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+The *computed* value for the `border-top-color` and the `border-bottom-color`
+switches from `rgba(0, 0, 0, 0.847)` (blackish) to `rgba(255, 255, 255, 0.847)` (whitish),
+since the user-agent updates `ButtonFace` dynamically based on the color scheme.
+The same applies for the `<button>`'s `color`
```
Comment (jpmedley) [inline src/site/content/en/blog/color-scheme/index.md]:
```suggestion
The same applies for the `<button>` element's `color`
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+
+You can see the effects of `color-scheme` applied to a large number of HTML elements
+in the demo embedded below, or
+[directly on Glitch](https://color-scheme-demo.glitch.me/).
+Remember that you need to switch your operating system to dark mode to enjoy this demo.
+
+The demo on purpose shows the WCAG&nbsp;AA and WCAG&nbsp;AAA
+[violation](https://webaim.org/resources/contrastchecker/?fcolor=0000EE&bcolor=000000)
+with the link colors that was mentioned in the
+[warning above](#using-color-scheme-in-practice).
```
Comment (jpmedley) [inline src/site/content/en/blog/color-scheme/index.md]:
```suggestion
The demo deliberately uses the WCAG&nbsp;AA and WCAG&nbsp;AAA
[violation](https://webaim.org/resources/contrastchecker/?fcolor=0000EE&bcolor=000000)
with the link colors mentioned in the
[warning above](#using-color-scheme-in-practice).
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 12]
```diff
@@ -0,0 +1,360 @@
+---
+title: The `color-scheme` CSS property and the corresponding meta tag
```
Comment (tomayac) [inline src/site/content/en/blog/color-scheme/index.md:2]:
```suggestion
title: Improved dark mode with the `color-scheme` CSS property and the corresponding meta tag
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+subhead: |
+  The `color-scheme` CSS property and the corresponding meta tag
```
[the remark this answers is not in the document]
Comment (tomayac) [reply src/site/content/en/blog/color-scheme/index.md:4]: It does.

File: src/site/content/en/blog/color-scheme/index.md [hunk 14]
```diff
@@ -0,0 +1,361 @@
...
+  allow developers to opt in their pages to theme-specific defaults of the user-agent stylesheet.
+authors:
+  - thomassteiner
+date: 2020-03-27
+hero: hero.jpg
+alt: Pigeons on a wall with a sharp black and white contrast in the background.
+description: |
+  The color-scheme CSS property and the corresponding meta tag
```
[the remark this answers is not in the document]
Comment (tomayac) [reply src/site/content/en/blog/color-scheme/index.md:12]: This is for the description meta tag, so here it doesn't show.

File: src/site/content/en/blog/color-scheme/index.md [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+[Firefox's](https://dxr.mozilla.org/mozilla-central/source/layout/style/res/html.css) or
+[Safari's](https://trac.webkit.org/browser/trunk/Source/WebCore/css/html.css) (and WebKit's).
+Typically, UA stylesheets agree on the majority of things.
+For example, they all make links blue, general text black, and background color white,
+but there are also important (and sometimes annoying) differences,
+for instance, how they style form controls.
+
+Have a closer look at
+[WebKit's UA stylesheet](https://trac.webkit.org/browser/trunk/Source/WebCore/css/html.css)
+and what it does regarding dark mode (do a full text search for "dark" in the code).
```
Comment (tomayac) [inline src/site/content/en/blog/color-scheme/index.md:62]:
```suggestion
and what it does regarding dark mode.
(Do a full text search for "dark" in the stylesheet.)
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 16]
```diff
@@ -0,0 +1,361 @@
...
+
+Have a closer look at
+[WebKit's UA stylesheet](https://trac.webkit.org/browser/trunk/Source/WebCore/css/html.css)
+and what it does regarding dark mode.
+(Do a full text search for "dark" in the stylesheet.)
+The default provided by the stylesheet changes based on whether dark mode is on or off.
+To illustrate this, here is one such CSS rule (using the
+[`:matches`](https://css-tricks.com/almanac/selectors/m/matches/)
+pseudo class and WebKit-internal variables like `-apple-system-control-background`,
+as well as the WebKit-internal preprocessor directive `#if defined`):
```
Comment (tomayac) [inline src/site/content/en/blog/color-scheme/index.md:68]:
```suggestion
To illustrate this, here is one such CSS rule using the
[`:matches`](https://css-tricks.com/almanac/selectors/m/matches/)
pseudo class and WebKit-internal variables like `-apple-system-control-background`,
as well as the WebKit-internal preprocessor directive `#if defined`:
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 17]
```diff
@@ -0,0 +1,361 @@
...
+is for the background of application content or documents,
+whereas
+[`CanvasText`](https://drafts.csswg.org/css-color/#valdef-system-color-canvastext)
+is for text in application content or documents.
+The two go together and should not be used in isolation.
+
+UA stylesheets can make use of their own proprietary, or the standardized semantic system colors,
+to determine how HTML elements should be rendered by default.
+If the operating system is set to dark mode or uses a dark theme,
+`CanvasText` (or `text` respectively) would be conditionally set to whitish,
```
[the remark this answers is not in the document]
Comment (tomayac) [reply src/site/content/en/blog/color-scheme/index.md:104]: `text` is the WebKit-internal way to mean `CanvasText`.

File: src/site/content/en/blog/color-scheme/index.md [hunk 18]
```diff
@@ -0,0 +1,361 @@
...
+is for the background of application content or documents,
+whereas
+[`CanvasText`](https://drafts.csswg.org/css-color/#valdef-system-color-canvastext)
+is for text in application content or documents.
+The two go together and should not be used in isolation.
+
+UA stylesheets can use their own proprietary, or the standardized semantic system colors,
+to determine how HTML elements should be rendered by default.
+If the operating system is set to dark mode or uses a dark theme,
+`CanvasText` (or `text` respectively) would be conditionally set to whitish,
```
Comment (tomayac) [inline src/site/content/en/blog/color-scheme/index.md:104]:
```suggestion
`CanvasText` (or `text` respectively) would be conditionally set to white,
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 19]
```diff
@@ -0,0 +1,361 @@
...
+whereas
+[`CanvasText`](https://drafts.csswg.org/css-color/#valdef-system-color-canvastext)
+is for text in application content or documents.
+The two go together and should not be used in isolation.
+
+UA stylesheets can use their own proprietary, or the standardized semantic system colors,
+to determine how HTML elements should be rendered by default.
+If the operating system is set to dark mode or uses a dark theme,
+`CanvasText` (or `text` respectively) would be conditionally set to white,
+and `Canvas` (or `-apple-system-control-background`) would be set to blackish.
```
Comment (tomayac) [inline src/site/content/en/blog/color-scheme/index.md:105]:
```suggestion
and `Canvas` (or `-apple-system-control-background`) would be set to black.
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 20]
```diff
@@ -0,0 +1,361 @@
...
+is for the background of application content or documents,
+whereas
+[`CanvasText`](https://drafts.csswg.org/css-color/#valdef-system-color-canvastext)
+is for text in application content or documents.
+The two go together and should not be used in isolation.
+
+UA stylesheets can use their own proprietary, or the standardized semantic system colors,
+to determine how HTML elements should be rendered by default.
+If the operating system is set to dark mode or uses a dark theme,
+`CanvasText` (or `text` respectively) would be conditionally set to white,
```
Comment (tomayac) [inline src/site/content/en/blog/color-scheme/index.md:104]:
```suggestion
`CanvasText` (or `text`) would be conditionally set to white,
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 21]
```diff
@@ -0,0 +1,361 @@
...
+</figure>
+
+## Demo
+
+You can see the effects of `color-scheme` applied to a large number of HTML elements
+in the demo embedded below, or
+[directly on Glitch](https://color-scheme-demo.glitch.me/).
+Remember that you need to switch your operating system to dark mode to enjoy this demo.
+
+The demo deliberately uses the WCAG&nbsp;AA and WCAG&nbsp;AAA
```
Comment (tomayac) [inline src/site/content/en/blog/color-scheme/index.md:339]:
```suggestion
The demo deliberately shows the WCAG&nbsp;AA and WCAG&nbsp;AAA
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 22]
```diff
@@ -0,0 +1,361 @@
...
+draft: true
+---
+## Background
+
+### The `prefers-color-scheme` user preference media feature
+
+The
+[`prefers-color-scheme`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme)
+user preference media feature gives developers full control over their pages' appearances.
+If you're unfamiliar with it read my article
```
Comment (tomayac) [inline src/site/content/en/blog/color-scheme/index.md:32]:
```suggestion
If you are unfamiliar with it read my article
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 23]
```diff
@@ -0,0 +1,361 @@
...
+  color: CanvasText;
+  background-color: Canvas
+}
+```
+
+## The `color-scheme` CSS property
+
+The [CSS Color Adjustment Module Level&nbsp;1](https://drafts.csswg.org/css-color-adjust/)
+specification introduces a model and controls
+over automatic color adjustment by the user agent to handle user preferences,
```
Comment (tomayac) [inline src/site/content/en/blog/color-scheme/index.md:123]:
```suggestion
over automatic color adjustment by the user agent
with the objective of handling user preferences
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 24]
```diff
@@ -0,0 +1,361 @@
...
+  background-color: Canvas
+}
+```
+
+## The `color-scheme` CSS property
+
+The [CSS Color Adjustment Module Level&nbsp;1](https://drafts.csswg.org/css-color-adjust/)
+specification introduces a model and controls
+over automatic color adjustment by the user-agent to handle user preferences,
+such as dark mode, contrast adjustment, or specific desired color schemes.
```
[the remark this answers is not in the document]
Comment (tomayac) [reply src/site/content/en/blog/color-scheme/index.md:125]: It's straight from the spec, including the wrong comma. I've fixed it.

**Comment by lilles** [comment]
Thanks Thomas!

A couple of notes:

Do we have a crbug report for the link colors? We shouldn't break accessibility requirements for the default rendering of:

```
<meta name="color-scheme" content="dark>
<a href="link.html">Click me</a>

```

Since the UA color changes for system colors, forms, scrollbars etc. has not shipped yet. Did you plan to publish this before that ships? If so, it should have a note about that?

**Comment by tomayac** [comment]
> Do we have a crbug report for the link colors? We shouldn't break accessibility requirements for the default rendering of:
> 
> ```
> <meta name="color-scheme" content="dark>
> <a href="link.html">Click me</a>
> ```

I've opened https://crbug.com/1066811 and will open corresponding WebKit and Firefox bugs.

> Since the UA color changes for system colors, forms, scrollbars etc. has not shipped yet. Did you plan to publish this before that ships? If so, it should have a note about that?

Content on web.dev is not tied to Chrome, we write about general web features. Ideally it'd work in Chrome at the time of publication, but it doesn't have to.

File: src/site/content/en/blog/color-scheme/index.md [hunk 25]
```diff
@@ -0,0 +1,376 @@
...
+  [fails](https://webaim.org/resources/contrastchecker/?fcolor=0000EE&bcolor=000000)
+  both the WCAG&nbsp;AA as well as the WCAG&nbsp;AAA
+  [requirements](https://www.w3.org/WAI/WCAG21/Understanding/conformance#levels).
+
+  I have opened bugs for [Chrome](https://crbug.com/1066811),
+  [WebKit](https://bugs.webkit.org/show_bug.cgi?id=209851), and
+  [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1626560)
+  as well as a [meta issue in the HTML Standard](https://github.com/whatwg/html/issues/5426)
+  to get this fixed.
+{% endAside %}
```
Comment (tomayac) [inline src/site/content/en/blog/color-scheme/index.md:221]:
```suggestion
While for absolute baseline pages no additional CSS rules are necessary,
in the general case you should always combine `color-scheme` with `prefers-color-scheme`.

For example, the proprietary WebKit CSS color `-webkit-link`, used by WebKit and Chrome
for the classic link blue `rgb(0,0,238)`,
has an insufficient contrast ratio of 2.23:1 on a black background and
[fails](https://webaim.org/resources/contrastchecker/?fcolor=0000EE&bcolor=000000)
both the WCAG&nbsp;AA as well as the WCAG&nbsp;AAA
[requirements](https://www.w3.org/WAI/WCAG21/Understanding/conformance#levels).

I have opened bugs for [Chrome](https://crbug.com/1066811),
[WebKit](https://bugs.webkit.org/show_bug.cgi?id=209851), and
[Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1626560)
as well as a [meta issue in the HTML Standard](https://github.com/whatwg/html/issues/5426)
to get this fixed.
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 26]
```diff
@@ -0,0 +1,374 @@
...
+## Using color-scheme in practice
+
+Since both the meta tag and the CSS property (if applied to the `:root` element)
+eventually result in the same behavior, I always recommend specifying the color scheme
+via the meta tag, so the browser can adopt to the preferred scheme faster.
+
+While for absolute baseline pages no additional CSS rules are necessary,
+in the general case you should always combine `color-scheme` with `prefers-color-scheme`.
+
+For example, the proprietary WebKit CSS color `-webkit-link`, used by WebKit and Chrome
```
Comment (tomayac) [inline src/site/content/en/blog/color-scheme/index.md:208]:
```suggestion
For example, the proprietary WebKit CSS color `-webkit-link`, used by WebKit and Chrome
```

File: src/site/content/en/blog/color-scheme/index.md [hunk 27]
```diff
@@ -0,0 +1,373 @@
...
+
+```html
+<!--
+  The page supports both dark and light color schemes,
+  and the page author prefers dark.
+-->
+<meta name="color-scheme" content="dark light">
+```
+
+## Using color-scheme in practice
```
Comment (tomayac) [inline src/site/content/en/blog/color-scheme/index.md:199]:
```suggestion
## Combining color-scheme and prefers-color-scheme
```

**Comment by kaycebasques** [comment]
Thanks for the great content @tomayac. Notes:

* The `draft: true` field is set in the front matter.  I'll assume that was intentional and that you want to make this live on a later date. If not feel free to open another PR or just ping me on Chat and I can push up a commit.
* We have the ability to embed Glitches into our pages. https://web.dev/handbook/web-dev-components/#glitches

**Comment by tomayac** [comment]
@kaycebasques Thanks, Kayce, for the tweaks.

> The draft: true field is set in the front matter. I'll assume that was intentional and that you want to make this live on a later date. If not feel free to open another PR or just ping me on Chat and I can push up a commit.

I've opened https://github.com/GoogleChrome/web.dev/pull/2532 to push it live. Are there comms from the team who should flip the `draft` bit (also see https://github.com/GoogleChrome/web.dev/issues/2519)? I was under the impression the merger flips the bit.

> We have the ability to embed Glitches into our pages. https://web.dev/handbook/web-dev-components/#glitches

Well aware, just in this case the embed doesn't work to convey the message due to iframe styling issues (https://github.com/w3c/csswg-drafts/issues/4772).

> The subhead is more descriptive IMO than the current title. After reading the current subhead (The color-scheme CSS property and the corresponding meta tag allow developers to opt their pages in to the theme-specific defaults of the user agent stylesheet.) I know exactly what I'm getting. Whereas with the current title (Improved dark mode with the color-scheme CSS property and the corresponding meta tag) I don't have much of a clue what the article actually discusses. I think it's important for the title to be more descriptive because that's sometimes the only thing that people see. I suggest swapping the title and subhead (of course you'll have to condense down the content in the subhead if that stuff becomes the title).

I have tweaked the title a bit https://github.com/GoogleChrome/web.dev/pull/2532, the "clickbait" is *improved dark mode default styling*, I think that works.

**Pull request merged**